doc/design/docker-cp.md

```diff
@@ -0,0 +1,144 @@
+
+```
+/contrainer/filesystem?ID= PUT
+
+query param:
+
+* ID - this is the target container
+
+Request Model : 
+
+* path - container filesystem target for the copy operation
+
+Response Model :
+
+* stream - tar stream of the data to be moved to the container
+
+* overwriteDir - flag for whether or not we are directed to overwrite existing directories 
+
+OR 
+
+* Error - error response from the portlayer
+
+```
+
+For a copy *from* the filesystem of a container we will do a `GET` operation. 
+
+```
+/container/filesystem?ID=
+
+Query Param: 
+
+* ID - this is the target container
+
+Request Model : 
+
+* path - target path for the copy operation
+
+* dataStream - tar stream of the data to be copies into the filesystem of the container
+
+OR 
+
+* Error - error response from the portlayer
+
+```
+
+### Personality design
+
+The personality has two endpoints that are called depending on the to/from situation above. Each of those endpoints will call their respective portlayer counterparts. There will be several checks during this process. 
+
+Error codes expected by docker : 
+
+```
+200 
+
+400
+
+403 - in the event of a readonly filesystem
+
+404
+
+500
+
+```
+
+Because we need to prevent a poweroff(unless forced) of a container during a copy, we will pass the ID directly to the portlayer rather than fetching a handle. 
+
+Docker does not do much in the way of string checking as far as the path is concerned. This is likely not a big concern for the personality as the actual copy commmand will fail if a bad path string is provided. 
+
+The personality code should be fairly straight forward, the goal is to call the portlayer for the copy while passing the tar stream either to the portlayer or from it. The best way to do this is to hook the writiers together using a PipeReader and then shepherding the data in the direction it needs to go. 
+
+## Portlayer Design
+
+ There will be two situations to be concerned with in the portlayer. When the container is *on* and when the container is *off*.
+ 
+ 
+### Off Container Scenario
+ 
+ This is the case where guest tools is not needed since the guest in off. This means that all action will need to occur on the VCH. There are several concerns and many of them are listed in this ticket [here](https://github.com/vmware/vic/issues/4105)
+ 
+ The most concern surrounding this piece of work will be surrounding VCH restart and confirming that the disk is still left in a good state if a copy was occuring mid restart. Additionally we need to block a container from starting when we have the r/w layer mounted. While that container is blocked from being started we might look at the possibility of returning a 500 error for a start invocation on a blocked container. We must also block other copy operations until the current one is successful, one major reason is because we cannot do any kind of sanity checking on available space with concurrent writes to the same disk. 
```

Comment:
We can do just as much sanity checking with concurrent writes as we can with serialized writes. I see no reason to prevent concurrent copy to/from _apart from the fact that we cannot provide stable semantics as to what occurs_. The same issues apply to concurrent copy with normal `cp` command and a local filesystem.

We should be returning "409 Conflict" rather than 500. Longer term we should probably allow configuration of whether to block in these circumstances or return an error - blocking is preferable in many automation scenarios as it prevents the need for the caller to differentiate errors that will resolve themselves in time.